### PR TITLE
Add macOS support

### PR DESCRIPTION
**Pull request by wydwww**
Tests passed on macOS(local and spark local) and ubuntu(local and spark local & cluster).
On ubuntu: same with before.
On macOS: need a full build, first install Intel Parallel Studio XE, then `bash make-dist.sh -P full-build -P mac`.
#340

### CODE REVIEW

**Comment by linbojin** [comment]
There is free Intel Parallel Studio XE for macOS?

**Comment by wydwww** [comment]
I'm not sure... You can try the trial version for now. Maybe we'll release a partial-pre-build version to avoid full build on mac later. @linbojin

**Comment by linbojin** [comment]
Thx, @wydwww

**Comment by jason-dai** [comment]
We plan to release jars for native lib code for both Linux and macOS afterwards, so that users will just need to do a normal build for macOS.

File: native/jni/src/main/java/com/intel/analytics/bigdl/mkl/MKL.java [hunk 1]
```diff
@@ -17,13 +17,24 @@
...
     static {
         try {
-            tmpFile = extract("libiomp5.so");
-            System.load(tmpFile.getAbsolutePath());
-            tmpFile.delete(); // delete so temp file after loaded
-            tmpFile = extract("libjmkl.so");
-            System.load(tmpFile.getAbsolutePath());
-            tmpFile.delete(); // delete so temp file after loaded
-            isLoaded = true;
+            if (System.getProperty("os.name").toLowerCase().contains("mac")) {
```
Comment (yiheng) [inline native/jni/src/main/java/com/intel/analytics/bigdl/mkl/MKL.java]: too much duplicate code here?  I think you just need set the file name based on os.name

File: native/pom.xml [hunk 2]
```diff
@@ -16,6 +16,31 @@
...
+    <profiles>
+        <profile>
+            <id>default</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
+            </activation>
+            <properties>
+                <compilerStartOptionDirectory>-I ${JAVA_HOME}/include/linux/</compilerStartOptionDirectory>
+                <linkerStartOptionDirectory>-I ${JAVA_HOME}/include/linux/</linkerStartOptionDirectory>
+                <packagingDirectory>so</packagingDirectory>
```
Comment (yiheng) [inline native/pom.xml]: packagingDirectory -> dynamicLibType?

File: native/pom.xml [hunk 3]
```diff
@@ -16,6 +16,31 @@
...
+                <includeDirectory>libiomp5.so</includeDirectory>
```
Comment (yiheng) [inline native/pom.xml]: includeDirectory -> iompLib?

File: native/pom.xml [hunk 4]
```diff
@@ -16,6 +16,31 @@
...
+                <typeDirectory>so</typeDirectory>
```
Comment (yiheng) [inline native/pom.xml]: reuse dynamicLibType?

File: native/pom.xml [hunk 5]
```diff
@@ -16,6 +16,31 @@
...
+                <destFileNameDirectory>libjmkl.so</destFileNameDirectory>
```
Comment (yiheng) [inline native/pom.xml]: destFileNameDirectory -> destFileName?

BTW, why you always have a Directory suffix...

File: native/jni/src/main/resources/intel64/libiomp5.dylib [hunk 6]
```diff
@@ -0,0 +1 @@
+/opt/intel/lib/libiomp5.dylib
```
Comment (yiheng) [inline native/jni/src/main/resources/intel64/libiomp5.dylib]: what's this for?

**Comment by yiheng** [comment]
I think in the script, you should change the artifactId of jni module instead of native module? We'll upload two jni jar to the public repo. jni(default for linux x64) and jni-mac. So user on their mac only need to type mvn package -P mac, they will download the jni-mac jar instead of jni jar. So they needn't install Parallel studio or compile native code.

**Comment by yiheng** [comment]
In this case, I think we need a script to change the artifiactId but a just another module jni-mac. And we can active the mac profile base on os. No need user input -P mac. See http://maven.apache.org/guides/introduction/introduction-to-profiles.html for how to do it.

**Comment by jason-dai** [comment]
I don't think using script for this type of tricks is needed; just add a parameter and then the user can manage it explicitly.

File: native/jni/src/main/resources/intel64/libiomp5.dylib (hunk 6 above)
Comment (wydwww) [inline native/jni/src/main/resources/intel64/libiomp5.dylib]: Soft link to libiomp5.dylib of intel parallel studio, just like libiomp5.so

Comment (yiheng) [inline native/jni/src/main/resources/intel64/libiomp5.dylib]: I think it's not a good way to handle this, as the path is hard coded..

File: native/jni/src/main/java/com/intel/analytics/bigdl/mkl/MKL.java [hunk 7]
```diff
@@ -17,24 +17,20 @@
 
     static {
         try {
+            String fileName1 = "libiomp5.so";
```
Comment (yiheng) [inline native/jni/src/main/java/com/intel/analytics/bigdl/mkl/MKL.java]: don't use fileName1 and fileName2 as variable name. It's too confusing

File: native/pom.xml [hunk 8]
```diff
@@ -25,21 +25,19 @@
             <properties>
                 <compilerStartOptionDirectory>-I ${JAVA_HOME}/include/linux/</compilerStartOptionDirectory>
                 <linkerStartOptionDirectory>-I ${JAVA_HOME}/include/linux/</linkerStartOptionDirectory>
-                <packagingDirectory>so</packagingDirectory>
-                <includeDirectory>libiomp5.so</includeDirectory>
-                <typeDirectory>so</typeDirectory>
-                <destFileNameDirectory>libjmkl.so</destFileNameDirectory>
+                <dynamicLibType>so</dynamicLibType>
```
Comment (yiheng) [inline native/pom.xml]: I don't think the default profile is necessary.

File: dl/pom.xml [hunk 9]
```diff
@@ -244,6 +244,7 @@
             <properties>
                 <spark-version.version>1.5-plus</spark-version.version>
                 <spark.version>1.5.1</spark.version>
+                <mkl-java-os-version>mkl-java</mkl-java-os-version>
```
Comment (yiheng) [inline dl/pom.xml]: move this property to bigdl-parent pom.xml

File: dl/pom.xml [hunk 10]
```diff
@@ -259,5 +260,11 @@
                 <spark-scope>compile</spark-scope>
             </properties>
         </profile>
+        <profile>
```
Comment (yiheng) [inline dl/pom.xml]: move this profile to bigdl-parent pom.xml

File: scripts/change-os.sh [hunk 11]
```diff
@@ -58,6 +58,6 @@ sed_i() {
 
 export -f sed_i
 
-BASEDIR=$(dirname $0)/../native/mkl
+BASEDIR=$(dirname $0)/../native/jni
```
Comment (yiheng) [inline scripts/change-os.sh]: we don't need this script

File: native/jni/pom.xml [hunk 12]
```diff
@@ -10,7 +10,7 @@
     <modelVersion>4.0.0</modelVersion>
 
     <groupId>com.intel.analytics.bigdl.native</groupId>
-    <artifactId>mkl-java</artifactId>
+    <artifactId>mkl-java-mac</artifactId>
```
Comment (yiheng) [inline native/jni/pom.xml]: you should not change it.

File: scripts/change-os.sh (hunk 11 above)
Comment (yiheng) [inline scripts/change-os.sh]: create a new sub-module project

File: native/pom.xml (hunk 8 above)
Comment (wydwww) [inline native/pom.xml]: You mean choose the properties automatically by OS?

Comment (wydwww) [inline native/pom.xml]: You mean choose the properties automatically by OS? @yiheng

Comment (yiheng) [inline native/pom.xml]: no. You needn't a default profile to make the property as default

Comment (wydwww) [inline native/pom.xml]: Because of the `activeByDefault` tag? So I just need to change `default` to `linux`? @yiheng

Comment (yiheng) [inline native/pom.xml]: No. You needn't a profile to make default value of property.

**Comment by linbojin** [comment]
I think you may need to rebase your long commits...

File: pom.xml [hunk 13]
```diff
@@ -514,9 +516,21 @@
             <id>full-build</id>
             <modules>
                 <module>native</module>
+                <module>dl</module>
```
Comment (yiheng) [inline pom.xml]: don't add this back

File: native/pom.xml [hunk 14]
```diff
@@ -26,6 +26,11 @@
     <profiles>
         <profile>
             <id>mac</id>
+            <activation>
```
Comment (jason-dai) [inline native/pom.xml]: I don't think we want to do this automatically; instead, have the user to explicitly enable this when needed.

Comment (wydwww) [inline native/pom.xml]: Fixed. Without `-P mac` it will build the linux jar on macOS. @jason-dai

**Comment by wydwww** [comment]
Successfully build on shiqing's mac. @yiheng

**Comment by yiheng** [comment]
test looks good. merged

**Comment by yiheng** [comment]
The pre-build package has been uploaded to sonatype

**Pull request merged**